src/resources/js/test/loginForm.test.tsx

```diff
@@ -1,27 +1,136 @@
 import { render, screen } from '@testing-library/react';
-import { describe, it } from 'vitest';
+import userEvent from '@testing-library/user-event';
+import { ResponseError, Token, User } from 'bmlt-root-server-client';
+import { describe, test, vi } from 'vitest';
 
-import { strings } from '../localization';
-import LoginForm from '../partials/forms/LoginForm';
-import { provideTheme } from './utils/provideTheme';
+import App from '../App';
+import ApiClientWrapper from '../RootServerApi';
+
+// define a mock authToken that expires 1 hour from now
+// (the token uses PHP's time rather than Javascript's time, so seconds from the epoch instead of milliseconds)
+const now: number = Math.round(new Date().valueOf() / 1000);
+const mockedToken: Token = {
+  accessToken: 'mysteryString42',
+  expiresAt: now + 60 * 60,
+  tokenType: 'bearer',
+  userId: 1,
+};
+
+// for now just define a serveradmin user
+const mockUser: User = {
+  description: 'Main Server Administrator',
+  displayName: 'Server Administrator',
+  email: 'nobody@bmlt.app',
+  id: 1,
+  ownerId: '',
+  type: 'admin',
+  username: 'serveradmin',
+};
+
+// mocked access token
+let savedAccessToken: Token | null;
+
+function mockGetToken(): Token | null {
+  return savedAccessToken;
+}
+
+function mockSetToken(token: Token | null): void {
+  savedAccessToken = token;
+}
+
+async function mockGetUser(params: { userId: number }): Promise<User> {
+  if (params.userId == mockUser.id) {
+    return mockUser;
+  }
+  throw new Error('no user found with the given userId');
+}
+
+function mockIsLoggedIn(): boolean {
+  return Boolean(savedAccessToken);
+}
+
+async function mockAuthToken(authTokenRequest: { tokenCredentials: { username: string; password: string } }): Promise<Token> {
+  if (authTokenRequest.tokenCredentials.username === 'serveradmin' && authTokenRequest.tokenCredentials.password === 'good-password') {
+    return mockedToken;
+  } else {
+    const msg = '{ "message": "The provided credentials are incorrect." }';
+    const unicodeMsg = Uint8Array.from(Array.from(msg).map((x) => x.charCodeAt(0)));
+    const strm = new ReadableStream({
+      start(controller) {
+        controller.enqueue(unicodeMsg);
+        controller.close();
+      },
+    });
+    const r: Response = new Response(strm, { status: 401, statusText: 'Unauthorized' });
```

Comment:
I think you can remove the `: Response` type hint here, and just have:

`const r = new Response(strm, { status: 401, statusText: 'Unauthorized' });`

It should be able to infer the type.